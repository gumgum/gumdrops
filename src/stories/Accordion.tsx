import React from 'react';
import cx from 'classnames';

interface AccordionCommonProps {
    color?: 'dark' | 'white';
    size?: 'sm' | '';
}

type OmitColor<T> = Omit<T, 'color'>;

export interface AccordionItemProps
    extends AccordionCommonProps,
        OmitColor<React.LiHTMLAttributes<HTMLLIElement>> {
    isOpen?: boolean;
    isLocked?: boolean;
    label: React.ReactChild;
}

export interface AccordionProps
    extends AccordionCommonProps,
        OmitColor<React.HTMLAttributes<Element>> {
    initialAllOpen?: boolean;
}

export const Accordion: React.FC<AccordionProps> = ({
    children,
    color,
    size = '',
    className,
    initialAllOpen,
    ...otherProps
}) => {
    const rootClass = cx('gds-accordion', className, {
        [`gds-accordion--${color}`]: color
    });

    const newChildren = React.Children.map(children, child => {
        if (!React.isValidElement<AccordionItemProps>(child)) {
            return child;
        }

        return React.cloneElement<AccordionItemProps>(child, {
            color,
            size,
            isOpen: child.props.isOpen ? child.props.isOpen : initialAllOpen
        });
    });

    return (
        <div className={rootClass} {...otherProps}>
            <ul className="gds-accordion-list">{newChildren}</ul>
        </div>
    );
};
