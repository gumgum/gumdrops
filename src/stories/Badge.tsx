import React, { ReactChild, ReactChildren } from 'react';
import cx from 'classnames';
import { Colors } from 'types';

export interface BadgeProps extends React.HTMLAttributes<Element> {
    text?: JSX.Element[] | ReactChildren | ReactChild;
    color?: Colors;
    empty?: boolean;
}

const baseClass = 'gds-badge';

export const Badge: React.FC<BadgeProps> = ({
    color = 'default',
    text,
    empty,
    className,
    ...otherProps
}) => {
    const rootClass = cx(baseClass, className, {
        [`${baseClass}--${color}`]: color,
        [`${baseClass}--empty`]: empty
    });

    return (
        <span className={rootClass} {...otherProps}>
            {!empty && text}
        </span>
    );
};
