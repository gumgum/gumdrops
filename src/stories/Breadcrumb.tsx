import React from 'react';
import cx from 'classnames';

import { BreadcrumbMenu } from './BreadcrumbMenu';

const baseClass = 'gds-page-header__breadcrumbs-list-item';

const filterMenu = (pathname: string) => ({ path }): boolean =>
    path.charAt(0) !== ':' && !pathname.includes(path);

export interface BreadCrumb {
    title?: string;
    path: string;
    subpaths?: BreadCrumb[];
}

export type LinkComponent = React.FunctionComponent<{ to: string; className: string }>;

export interface BreadcrumbProps extends BreadCrumb, React.HTMLAttributes<HTMLOListElement> {
    linkComponent?: LinkComponent;
    pathname: string;
    isLast?: boolean;
    hideMenus: boolean;
}

export const Breadcrumb: React.FC<BreadcrumbProps> = ({
    linkComponent: LinkComponent,
    title,
    path,
    subpaths,
    pathname,
    hideMenus,
    isLast,
    className
}) => {
    const hasSubpaths = !hideMenus && subpaths && subpaths.length;
    const menu = hasSubpaths && subpaths.filter(filterMenu(pathname));
    const hasMenu = !!(menu && menu.length);
    const rootClass = cx(baseClass, className, {
        [`${baseClass}--has-menu`]: hasMenu
    });
    const displayTitle = title || path;

    return (
        <li className={rootClass} aria-label={displayTitle}>
            {hasMenu && (
                <BreadcrumbMenu
                    linkComponent={LinkComponent}
                    menu={menu}
                    path={path}
                    sectionTitle={displayTitle}
                />
            )}
            {isLast ? (
                displayTitle
            ) : (
                <LinkComponent className="gds-page-header__breadcrumbs-link" to={path}>
                    {displayTitle}
                </LinkComponent>
            )}
        </li>
    );
};
