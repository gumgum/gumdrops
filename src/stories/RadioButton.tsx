import React, { ReactChild, ReactChildren } from 'react';
import cx from 'classnames';
import { Colors } from 'types';

export interface RadioButtonProps extends React.InputHTMLAttributes<HTMLInputElement> {
    label?: ReactChild | ReactChildren | Element;
    color?: Colors;
    checked?: boolean;
}

const baseClass = 'gds-form-group';

export const RadioButton: React.FC<RadioButtonProps> = ({
    color = 'default',
    label,
    className,
    ...otherProps
}) => {
    const rootClass = cx(baseClass, `${baseClass}__radio-label`, className, {
        [`${baseClass}--${color}`]: color
    });

    return (
        <label className={rootClass}>
            <input className="gds-form-group__radio-input" {...otherProps} type="radio" />
            <span className="gds-form-group__radio-indicator" />
            {label}
        </label>
    );
};
