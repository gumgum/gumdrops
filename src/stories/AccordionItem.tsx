import React, { MouseEvent, KeyboardEvent, useState } from 'react';
import cx from 'classnames';
import { AccordionItemProps } from './Accordion';
import { keyCodes } from 'types';

type ToggleEvent = MouseEvent<HTMLLIElement> | KeyboardEvent<HTMLLIElement>;

export const AccordionItem: React.FC<AccordionItemProps> = ({
    isOpen: initialIsOpen,
    isLocked,
    label,
    size,
    color,
    className,
    children
}) => {
    const [isOpen, setIsOpen] = useState(initialIsOpen);

    const toggleOpen = (event: ToggleEvent): void => {
        event.stopPropagation(); // don't want nested accordions to propagate events
        event.preventDefault(); // prevents page scroll from space key
        const { type } = event;
        const { key } = event as KeyboardEvent;

        if (isLocked) return;
        const allowToggle =
            (type === 'keypress' && [keyCodes.space, keyCodes.enter].includes(key as keyCodes)) ||
            type === 'click';

        if (!allowToggle) return;

        setIsOpen(isOpen => !isOpen);
    };

    const rootClass = cx('gds-accordion__item', className, {
        [`gds-accordion__item--${color}`]: color,
        'gds-accordion__item--active': isOpen
    });

    const titleClass = cx('gds-accordion__item-title', {
        [`gds-accordion__item-title--${size}`]: size
    });

    const iconClass = cx('gds-accordion__item-icon', {
        [`gds-accordion__item-icon--${size}`]: size
    });

    const childClass = cx('gds-accordion__child-items', {
        [`gds-accordion__child-items--${size}`]: size
    });

    const newChildren = React.Children.map(children, child => {
        if (!React.isValidElement<AccordionItemProps>(child)) {
            return child;
        }

        return React.cloneElement<AccordionItemProps>(child, {
            color,
            size
        });
    });

    const regionId = `AccordionItem-region-${label}`;
    const labelId = `AccordionItem-label-${label}`;

    return (
        <li
            aria-pressed={isOpen}
            aria-controls={regionId}
            aria-expanded={isOpen}
            className={rootClass}
            onClick={toggleOpen}
            onKeyPress={toggleOpen}
            role="button"
            tabIndex={isLocked ? -1 : 0}>
            <h4 id={labelId} className={titleClass}>
                {label}
            </h4>
            {!isLocked && <span className={iconClass} />}
            <ul
                aria-labelledby={labelId}
                aria-hidden={!isOpen}
                className={childClass}
                id={regionId}
                role="region">
                {newChildren}
            </ul>
        </li>
    );
};
