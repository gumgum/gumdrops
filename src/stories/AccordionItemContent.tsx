import React from 'react';
import cx from 'classnames';
import { AccordionItemProps } from './Accordion';

export const AccordionItemContent: React.FC<Omit<AccordionItemProps, 'label'>> = ({
    className,
    color,
    size,
    children,
    ...otherProps
}) => {
    const stopClick: React.MouseEventHandler = e => {
        e.stopPropagation();
    };

    const rootClass = cx('gds-accordion__child-item', className, {
        [`gds-accordion__child-item--${color}`]: color
    });

    const titleClass = cx('gds-accordion__child-item-title', {
        [`gds-accordion__child-item-title--${size}`]: size
    });

    return (
        <li className={rootClass} onClick={stopClick} {...otherProps}>
            <h4 className={titleClass}>{children}</h4>
        </li>
    );
};
