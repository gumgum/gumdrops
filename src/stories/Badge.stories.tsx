import React from 'react';
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Story, Meta } from '@storybook/react/types-6-0';

import { Badge, BadgeProps } from './Badge';
import { Colors } from '../types';

export default {
    title: 'Badge',
    component: Badge
} as Meta;

const Template: Story<BadgeProps> = args => <Badge {...args} />;

export const Primary = Template.bind({});
Primary.args = {
    color: Colors.primary,
    text: 'Primary badge'
};

export const Warning = Template.bind({});
Warning.args = {
    color: Colors.warning,
    text: 'Warning Badge'
};

export const Empty = Template.bind({});
Empty.args = {
    color: Colors.info,
    empty: true
};
