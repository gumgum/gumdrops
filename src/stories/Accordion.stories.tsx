import React from 'react';
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Story, Meta } from '@storybook/react/types-6-0';

import { Accordion, AccordionProps } from './Accordion';
import { AccordionItem } from './AccordionItem';
import { AccordionItemContent } from './AccordionItemContent';

export default {
    title: 'Accordion',
    component: Accordion
} as Meta;

const Template: Story<AccordionProps> = args => (
    <Accordion {...args}>
        <AccordionItem label="I'm locked and open" isLocked isOpen>
            <AccordionItemContent>Locked Content</AccordionItemContent>
        </AccordionItem>
        <AccordionItem label="Item 1">
            <AccordionItemContent>Content 1</AccordionItemContent>
        </AccordionItem>
        <AccordionItem label="Item 2">
            <AccordionItemContent>Content 2</AccordionItemContent>
        </AccordionItem>
        <AccordionItem label="Item 3">
            <AccordionItemContent>Content 3</AccordionItemContent>
            <Accordion>
                <AccordionItem label="Nested Item">
                    <AccordionItemContent>Nested Content</AccordionItemContent>
                </AccordionItem>
            </Accordion>
        </AccordionItem>
    </Accordion>
);

export const Primary = Template.bind({});
Primary.args = {};
