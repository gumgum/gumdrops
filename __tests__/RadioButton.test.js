import React from 'react';
import { render, screen } from '@testing-library/react';
import user from '@testing-library/user-event';
import { RadioButton } from 'src/stories/RadioButton';

describe('RadioButton', () => {
    it('calls the given callback on click', () => {
        const mockOnChange = jest.fn();
        const { container } = render(<RadioButton onChange={mockOnChange} label="My Checkbox" />);
        const check = screen.getByRole('radio', { name: 'My Checkbox' });
        expect(check).not.toBeChecked();
        user.click(check);
        expect(mockOnChange).toHaveBeenCalledTimes(1);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders a value', () => {
        const mockOnChange = jest.fn();
        const { container } = render(
            <RadioButton onChange={mockOnChange} label="My Checkbox" checked={true} />
        );
        expect(screen.getByRole('radio', { name: 'My Checkbox' })).toBeChecked();
        expect(container.firstChild).toMatchSnapshot();
    });
});
