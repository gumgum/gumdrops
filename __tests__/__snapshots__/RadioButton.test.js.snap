// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton calls the given callback on click 1`] = `
<label
  class="gds-form-group gds-form-group__radio-label gds-form-group--default"
>
  <input
    class="gds-form-group__radio-input"
    type="radio"
  />
  <span
    class="gds-form-group__radio-indicator"
  />
  My Checkbox
</label>
`;

exports[`RadioButton renders a value 1`] = `
<label
  class="gds-form-group gds-form-group__radio-label gds-form-group--default"
>
  <input
    checked=""
    class="gds-form-group__radio-input"
    type="radio"
  />
  <span
    class="gds-form-group__radio-indicator"
  />
  My Checkbox
</label>
`;
